File: backend/src/models/chatMessage.ts
import mongoose, { Document, Model, Schema } from 'mongoose';

export type SenderType = 'manager' | 'user';

export interface ChatMessageDocument extends Document {
  conversationId: mongoose.Types.ObjectId;
  managerId: mongoose.Types.ObjectId;
  userKey: string; // provider:subject format
  senderType: SenderType; // who sent the message
  senderName?: string; // sender's display name
  senderPicture?: string; // sender's profile picture
  message: string;
  readByManager: boolean;
  readByUser: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const ChatMessageSchema = new Schema<ChatMessageDocument>(
  {
    conversationId: {
      type: Schema.Types.ObjectId,
      ref: 'Conversation',
      required: true,
      index: true
    },
    managerId: {
      type: Schema.Types.ObjectId,
      ref: 'Manager',
      required: true,
      index: true
    },
    userKey: {
      type: String,
      required: true,
      trim: true,
      index: true
    },
    senderType: {
      type: String,
      enum: ['manager', 'user'],
      required: true
    },
    senderName: { type: String, trim: true },
    senderPicture: { type: String, trim: true },
    message: {
      type: String,
      required: true,
      trim: true,
      maxlength: 5000
    },
    readByManager: { type: Boolean, default: false, index: true },
    readByUser: { type: Boolean, default: false, index: true },
  },
  { timestamps: true }
);

// Index for querying messages by conversation and time
ChatMessageSchema.index({ conversationId: 1, createdAt: -1 });

// Index for counting unread messages
ChatMessageSchema.index({ conversationId: 1, readByManager: 1 });
ChatMessageSchema.index({ conversationId: 1, readByUser: 1 });

export const ChatMessageModel: Model<ChatMessageDocument> =
  mongoose.models.ChatMessage ||
  mongoose.model<ChatMessageDocument>('ChatMessage', ChatMessageSchema);

File: lib/features/chat/domain/entities/chat_message.dart
class ChatMessage {
  const ChatMessage({
    required this.id,
    required this.conversationId,
    required this.senderType,
    required this.message,
    required this.createdAt,
    this.senderName,
    this.senderPicture,
    this.readByManager = false,
    this.readByUser = false,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['id'] as String,
      conversationId: json['conversationId'] as String,
      senderType: json['senderType'] as String == 'manager'
          ? SenderType.manager
          : SenderType.user,
      senderName: json['senderName'] as String?,
      senderPicture: json['senderPicture'] as String?,
      message: json['message'] as String,
      readByManager: json['readByManager'] as bool? ?? false,
      readByUser: json['readByUser'] as bool? ?? false,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  final String id;
  final String conversationId;
  final SenderType senderType;
  final String? senderName;
  final String? senderPicture;
  final String message;
  final bool readByManager;
  final bool readByUser;
  final DateTime createdAt;

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'conversationId': conversationId,
      'senderType': senderType == SenderType.manager ? 'manager' : 'user',
      'senderName': senderName,
      'senderPicture': senderPicture,
      'message': message,
      'readByManager': readByManager,
      'readByUser': readByUser,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

enum SenderType {
  manager,
  user,
}
